feature-BJS2-96441-become-a-mentor

KuaNySH88 opened the pull request.

KuaNySH88 committed:
feature-BJS2-96441-become-a-mentor

KuaNySH88 committed:
feature-BJS2-96441-become-a-mentor fix checkstlyle comments

KuaNySH88 committed:
feature-BJS2-96441-become-a-mentor fix checkstlyle comments

maxmerdock commented on src/main/java/school/faang/user_service/controller/mentorship/MentorshipController.java:
```diff
@@ -0,0 +1,41 @@
+import java.util.List;
+
+@RestController
+@RequestMapping("/api/mentorships")
```
Про /api. Все другие контроллеры должны будут то же в начале содержать /api.
Лучше пропиши его в **properties** в **application.yaml**:

server:
  servlet:
    context-path: /api

Тогда все запросы автоматом будут начинаться с /api. И не надо будет дублировать это во все классы

maxmerdock replied:
Сделай плиз как у меня. Это же замечание было у моей задачи
https://github.com/CorporationX/user_service/pull/3335#discussion_r2398706896

maxmerdock commented on src/main/java/school/faang/user_service/service/mentorship/MentorshipServiceImpl.java:
```diff
@@ -0,0 +1,86 @@
+    private final UserMapper userMapper;
+    private final UserContext userContext;
+
+    @Override
```
ты тестил через Postman этот запрос?
данные сохраняются в бд?
думаю, тут отсутствие транзакционности: Если добавление ментора в список менторов прошло успешно, но сохранение изменений в репозитории завершилось ошибкой, состояние системы может стать неконсистентным. То есть, **возможно**, нужно добавить аннотацию **@Transactional**

maxmerdock commented on src/main/java/school/faang/user_service/validation/mentorship/MentorshipValidationImpl.java:
```diff
@@ -0,0 +1,14 @@
+    @Override
+    public boolean canAddMentorship(long mentorId, long menteeId, BiFunction<Long, Long, Boolean> biFunction) {
+        return biFunction.apply(mentorId, menteeId);
+    }
```
Как-то сложно реализовано для читаемости кода)
может стоит метод попроще написать:
    @Override
    public boolean canAddMentorship(long mentorId, long menteeId) {
        return mentorId != menteeId;
    }

А в методе addMentorship поменять на:
    if (!mentorshipValidation.canAddMentorship(mentorId, menteeId)) {
        throw new DataValidationException("Mentor Id and Mentee id can't be equal");
    }

maxmerdock commented on src/main/java/school/faang/user_service/service/mentorship/MentorshipServiceImpl.java:
```diff
@@ -0,0 +1,86 @@
+        if (currentUserId != menteeId && currentUserId != mentorId) {
+            throw new ForbiddenException("You are not a participant in the deleting mentorship!");
+        }
+
```
я бы вынес эту проверку тоже в MentorshipValidationImpl:
    @Override
    public void validateUserIsParticipantInMentorship(long menteeId, long mentorId, long currentUserId) {
        if (currentUserId != menteeId && currentUserId != mentorId) {
            throw new ForbiddenException("You are not a participant in the deleting mentorship!");
        }
    }

А в deleteMentorship было бы так лаконичнее:
mentorshipValidation.validateUserIsParticipantInMentorship(menteeId, mentorId, currentUserId);

SnurnicynKonstantin commented on src/main/java/school/faang/user_service/entity/mentorship/Mentorship.java:
```diff
@@ -0,0 +1,39 @@
+@NoArgsConstructor
+@Entity
+@Table(name = "mentorship")
+public class Mentorship {
```
Эта entity не используется в коде. Значит ее можно удалить

KuaNySH88 replied:
удалил

SnurnicynKonstantin commented on src/main/java/school/faang/user_service/controller/mentorship/MentorshipController.java:
```diff
@@ -0,0 +1,41 @@
+import java.util.List;
+
+@RestController
+@RequestMapping("/api/mentorships")
```
В проперти файле есть настройка, которая говорит о том, что все url запросов к сервису должны начинаться с /api - https://github.com/CorporationX/user_service/blob/kraken-master-stream12/src/main/resources/application.yaml#L27

По этому здесь можно удалить `/api`. Если бы в проперти не выставляли такой настройки, то пришлось бы в каждом контроллере добавлять /api. Что увеличивает когнитивную нагрузку

KuaNySH88 replied:
в application.yml не нашел такую настройку, добавил и удалил из контроллера

SnurnicynKonstantin replied:
Я имел ввиду, что в примере есть настройка. Не в нашей ветке

SnurnicynKonstantin commented on src/main/java/school/faang/user_service/service/mentorship/MentorshipServiceImpl.java:
```diff
@@ -0,0 +1,86 @@
+    @Override
+    public List<UserDto> getMentees(long userId) {
+        User mentor = mentorshipRepository.getByIdOrThrow(userId);
+        return Optional.ofNullable(mentor.getMentees())
```
Можно сразу
```
        return mentor.getMentees()
                .stream()
                .map(userMapper::toUserDto)
                .toList();
```
getMentees() если пустой список вернет, то и стрим пустой то же вернет

KuaNySH88 replied:
переписал

SnurnicynKonstantin commented on src/main/java/school/faang/user_service/service/mentorship/MentorshipServiceImpl.java:
```diff
@@ -0,0 +1,86 @@
+    @Override
+    public List<UserDto> getMentors(long userId) {
+        User mentee = mentorshipRepository.getByIdOrThrow(userId);
+        return Optional.ofNullable(mentee.getMentors())
```
То же замечание, что и для getMentees

KuaNySH88 replied:
переписал

KuaNySH88 committed:
feature-BJS2-96441-become-a-mentor fix PR comments:
1 add config for a default controller url
2 delete a redundant entity
3 delete a redundant using Optional empty List

SnurnicynKonstantin commented on src/main/resources/application.yaml:
```diff
@@ -23,6 +23,8 @@ spring:
 server:
   port: 8080
+  servlet:
+    context-path: /api
```
У нас в мастере уже есть этот параметр - https://github.com/CorporationX/user_service/blob/kraken-master-stream12/src/main/resources/application.yaml#L27

Удали это изменение. Да замержим.

KuaNySH88 replied:
удалил

KuaNySH88 committed:
feature-BJS2-96441-become-a-mentor fix PR comments:
1 delete config for a default controller url

SnurnicynKonstantin reviewed (approved):
Реализовал классы согласно общепринятой структуре проектов в джава. Здорово!
